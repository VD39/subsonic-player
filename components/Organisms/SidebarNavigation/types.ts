export interface SidebarButtonProps {
  icon: string;
  text: string;
}
