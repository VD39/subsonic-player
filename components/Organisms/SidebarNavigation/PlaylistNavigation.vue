<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';

import NavigationItem from './Items/NavigationItem.vue';
import SubNavigationItem from './Items/SubNavigationItem.vue';

defineProps<{
  collapsed: boolean;
  playlists: Playlist[];
}>();

defineEmits<{
  addPlaylist: [];
}>();
</script>

<template>
  <ul class="mBM">
    <NavigationItem :collapsed="collapsed" title="Playlist">
      <template v-if="playlists.length">
        <SubNavigationItem
          v-for="playlist in playlists.slice(0, 5)"
          :key="playlist.id"
          :collapsed="collapsed"
          :icon="ICONS.playlist"
          :title="playlist.name"
          :to="`/playlist/${playlist.id}`"
        />
      </template>
    </NavigationItem>

    <li>
      <ButtonLink
        ref="addPlaylist"
        class="sidebarLink"
        full-width
        :icon="ICONS.playlistAdd"
        :show-text="!collapsed"
        title="Add playlist"
        @click="$emit('addPlaylist')"
      >
        Add playlist
      </ButtonLink>
    </li>
  </ul>
</template>
