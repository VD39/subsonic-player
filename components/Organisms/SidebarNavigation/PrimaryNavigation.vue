<script setup lang="ts">
import NavigationItem from './Items/NavigationItem.vue';
import SubNavigationItem from './Items/SubNavigationItem.vue';

defineProps<{
  collapsed: boolean;
}>();

const { drop } = useDragAndDrop();
</script>

<template>
  <ul class="mBM">
    <NavigationItem
      v-for="items in SIDEBAR_DESKTOP_NAVIGATION"
      :key="`navigation-${items.title}`"
      class="mBM"
      :collapsed
      :title="items.title"
    >
      <SubNavigationItem
        v-for="item in items.items"
        :key="`item-${item.title}`"
        :class="[
          {
            [DRAG_AND_DROP_CLASS_NAMES.isDroppable]: item.isDroppable,
          },
        ]"
        :collapsed
        :icon="item.icon"
        :title="item.title"
        :to="item.to"
        @drop="drop(QUEUE_ID, $event)"
      />
    </NavigationItem>
  </ul>
</template>
