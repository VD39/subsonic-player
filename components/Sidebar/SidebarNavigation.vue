<script setup lang="ts">
import SidebarItems from '@/components/Sidebar/SidebarItems.vue';
import IconButton from '@/components/Buttons/IconButton.vue';
import MusicLogo from '@/components/Logo/MusicLogo.vue';

const { collapsed, toggle } = useSidebar();

const buttonProps = computed<ButtonProps>(() =>
  collapsed.value
    ? {
        icon: 'PhTextIndent',
        text: 'Open Navigation',
      }
    : {
        icon: 'PhTextOutdent',
        text: 'Close Navigation',
      },
);
</script>

<template>
  <div :class="$style.sideBar">
    <div :class="$style.header">
      <MusicLogo :class="$style.logo" />

      <IconButton
        ref="toggleButton"
        :icon="buttonProps.icon"
        :title="buttonProps.text"
        @click="toggle"
      >
        {{ buttonProps.text }}
      </IconButton>
    </div>

    <nav :class="$style.navigationWrapper">
      <ul :class="$style.navigation">
        <SidebarItems
          v-for="(item, index) in SIDEBAR_NAVIGATION"
          :key="`navigation-${index}`"
          :title="item.title"
          :title-icon="item.icon"
          :title-to="item.to"
          :items="item.items"
        />
      </ul>
    </nav>
  </div>
</template>

<style module>
.sideBar {
  @mixin align-center;

  position: fixed;
  inset: auto 0 0;
  z-index: 1;
  width: 100%;
  height: var(--header-height);
  padding-top: 0;
  background-color: var(--background-color);
  border-top: 1px solid var(--border-color);
  transition: width var(--transition);

  @media (--tablet-up) {
    inset: 0 auto 0 0;
    display: unset;
    width: var(--sidebar-width);
    height: auto;
    min-height: 100vh;
    overflow-x: auto;
    border-top: none;
    border-right: 1px solid var(--border-color);
  }
}

.header {
  display: none;

  @media (--tablet-up) {
    @mixin align-center;

    justify-content: space-between;
    min-height: var(--header-height);
    padding: 0 var(--space-16) 0 var(--space-24);
    margin-bottom: var(--space-16);

    :global(.collapsed) & {
      justify-content: center;
      padding: 0 var(--space-24);
    }
  }
}

.logo {
  :global(.collapsed) & {
    @mixin visually-hidden;
  }
}

.navigationWrapper {
  width: 100%;
}

.navigation {
  @media (--mobile-only) {
    @mixin inner;

    display: flex;
    gap: var(--space-40);
    justify-content: space-between;
    font-size: var(--small-font-size);
  }
}
</style>
