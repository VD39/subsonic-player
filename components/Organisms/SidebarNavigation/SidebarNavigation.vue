<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import MusicLogo from '@/components/Molecules/MusicLogo.vue';

import PlaylistNavigation from './PlaylistNavigation.vue';
import PrimaryNavigation from './PrimaryNavigation.vue';

const { collapsed, toggle } = useSidebar();

const buttonProps = computed<ButtonProps>(() => ({
  icon: collapsed.value ? ICONS.sidebarCollapsed : ICONS.sidebarNotCollapsed,
  text: collapsed.value ? 'Open Navigation' : 'Close Navigation',
}));
</script>

<template>
  <div
    :class="[
      'centerItems',
      'column',
      $style.sidebarNavigation,
      {
        [$style.collapsed]: collapsed,
      },
    ]"
  >
    <div :class="['mBM', 'centerAll', 'spaceBetween', $style.header]">
      <MusicLogo :class="$style.musicLogo" />

      <ButtonLink
        ref="toggleButton"
        :icon="buttonProps.icon"
        :title="buttonProps.text"
        @click="toggle"
      >
        {{ buttonProps.text }}
      </ButtonLink>
    </div>

    <nav class="fullWidth">
      <PrimaryNavigation :collapsed="collapsed" />

      <PlaylistNavigation :collapsed="collapsed" />
    </nav>
  </div>
</template>

<style module>
.sidebarNavigation {
  position: fixed;
  inset: 0 auto var(--sidebar-bottom) 0;
  z-index: 9;
  width: var(--sidebar-width);
  padding-top: 0;
  overflow-x: auto;
  background-color: var(--background-color);
  border-right: 1px solid var(--border-color);
  transition: width var(--transition);
}

.header {
  --header-padding: 0 var(--space-16) 0 var(--space-24);

  min-height: var(--header-height);
  padding: var(--header-padding);

  .collapsed & {
    --header-padding: 0 var(--space-24);
  }
}

.musicLogo {
  .collapsed & {
    display: none;
  }
}
</style>
