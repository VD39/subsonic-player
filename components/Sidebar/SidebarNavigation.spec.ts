import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import SidebarNavigation from './SidebarNavigation.vue';

function factory(props = {}) {
  return mount(SidebarNavigation, {
    props: {
      ...props,
    },
  });
}

describe('SidebarNavigation', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('sets the correct width to wrapper', () => {
    expect(wrapper.attributes('style')).toBe('width: 16rem;');
  });

  it('sets the correct icon on toggle button', () => {
    expect(wrapper.findComponent({ ref: 'toggleButton' }).props('icon')).toBe(
      'angles-left',
    );
  });

  it('sets the correct title on toggle button', () => {
    expect(
      wrapper.findComponent({ ref: 'toggleButton' }).attributes('title'),
    ).toBe('Close Navigation');
  });

  describe('when toggle button is clicked', () => {
    beforeEach(async () => {
      wrapper.findComponent({ ref: 'toggleButton' }).vm.$emit('click');
      await wrapper.vm.$nextTick();
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('sets the correct width to wrapper', () => {
      expect(wrapper.attributes('style')).toBe('width: 5rem;');
    });

    it('sets the correct icon on toggle button', () => {
      expect(wrapper.findComponent({ ref: 'toggleButton' }).props('icon')).toBe(
        'angles-right',
      );
    });

    it('sets the correct title on toggle button', () => {
      expect(
        wrapper.findComponent({ ref: 'toggleButton' }).attributes('title'),
      ).toBe('Open Navigation');
    });
  });
});
