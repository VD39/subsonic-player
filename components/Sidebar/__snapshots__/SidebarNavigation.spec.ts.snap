// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarNavigation > matches the snapshot 1`] = `
"<div class="sideBar">
  <div class="header"><a class="iconButton logo logo" title="Home">
      <phwavesine class="icon" aria-hidden="true" size="32" weight="duotone"></phwavesine><span class="text visually-hidden"> Home </span>
    </a><button class="iconButton" type="button" title="Close Navigation">
      <phtextoutdent class="icon" aria-hidden="true" size="24" weight="duotone"></phtextoutdent><span class="text visually-hidden">Close Navigation</span>
    </button></div>
  <nav class="navigationWrapper">
    <ul class="navigation">
      <li class="sidebarItem"><a class="titleLink" title="Browse">
          <phcompass class="titleIcon" aria-hidden="true" weight="duotone" size="32"></phcompass>
          <h4 class="title">Browse</h4>
        </a>
        <ul class="items">
          <li><a exactactiveclass="current" class="iconButton link link" title="Discover">
              <phcompass class="icon" aria-hidden="true" size="24" weight="duotone"></phcompass><span class="text">Discover</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Music">
              <phmusicnotes class="icon" aria-hidden="true" size="24" weight="duotone"></phmusicnotes><span class="text">Music</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Radio Stations">
              <phradio class="icon" aria-hidden="true" size="24" weight="duotone"></phradio><span class="text">Radio Stations</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Podcasts">
              <phapplepodcastslogo class="icon" aria-hidden="true" size="24" weight="duotone"></phapplepodcastslogo><span class="text">Podcasts</span>
            </a>
            <!--v-if-->
          </li>
        </ul>
      </li>
      <li class="sidebarItem"><a class="titleLink" title="Your library">
          <phmusicnotes class="titleIcon" aria-hidden="true" weight="duotone" size="32"></phmusicnotes>
          <h4 class="title">Your library</h4>
        </a>
        <ul class="items">
          <li><a exactactiveclass="current" class="iconButton link link" title="Albums">
              <phvinylrecord class="icon" aria-hidden="true" size="24" weight="duotone"></phvinylrecord><span class="text">Albums</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Artists">
              <phusersthree class="icon" aria-hidden="true" size="24" weight="duotone"></phusersthree><span class="text">Artists</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Playlists">
              <phplaylist class="icon" aria-hidden="true" size="24" weight="duotone"></phplaylist><span class="text">Playlists</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Favourites">
              <phheart class="icon" aria-hidden="true" size="24" weight="duotone"></phheart><span class="text">Favourites</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Genres">
              <phwaveform class="icon" aria-hidden="true" size="24" weight="duotone"></phwaveform><span class="text">Genres</span>
            </a>
            <!--v-if-->
          </li>
        </ul>
      </li>
      <li class="sidebarItem"><a class="titleLink" title="Playlists">
          <phplaylist class="titleIcon" aria-hidden="true" weight="duotone" size="32"></phplaylist>
          <h4 class="title">Playlists</h4>
        </a>
        <ul class="items">
          <li><a exactactiveclass="current" class="iconButton link link" title="Random">
              <phqueue class="icon" aria-hidden="true" size="24" weight="duotone"></phqueue><span class="text">Random</span>
            </a>
            <!--v-if-->
          </li>
          <li>
            <!--v-if-->
            <!--v-if-->
          </li>
          <li>
            <!--v-if--><button class="iconButton fullWidth link link" type="button" title="Add playlist" data-test-id="item-component">
              <phlistplus class="icon" aria-hidden="true" size="24" weight="duotone"></phlistplus><span class="text"> Add playlist </span>
            </button>
          </li>
        </ul>
      </li>
    </ul>
  </nav>
</div>"
`;

exports[`SidebarNavigation > when toggle button is clicked > matches the snapshot 1`] = `
"<div class="sideBar">
  <div class="header"><a class="iconButton logo logo" title="Home">
      <phwavesine class="icon" aria-hidden="true" size="32" weight="duotone"></phwavesine><span class="text visually-hidden"> Home </span>
    </a><button class="iconButton" type="button" title="Open Navigation">
      <phtextindent class="icon" aria-hidden="true" size="24" weight="duotone"></phtextindent><span class="text visually-hidden">Open Navigation</span>
    </button></div>
  <nav class="navigationWrapper">
    <ul class="navigation">
      <li class="sidebarItem"><a class="titleLink" title="Browse">
          <phcompass class="titleIcon" aria-hidden="true" weight="duotone" size="32"></phcompass>
          <h4 class="title">Browse</h4>
        </a>
        <ul class="items">
          <li><a exactactiveclass="current" class="iconButton link link" title="Discover">
              <phcompass class="icon" aria-hidden="true" size="24" weight="duotone"></phcompass><span class="text">Discover</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Music">
              <phmusicnotes class="icon" aria-hidden="true" size="24" weight="duotone"></phmusicnotes><span class="text">Music</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Radio Stations">
              <phradio class="icon" aria-hidden="true" size="24" weight="duotone"></phradio><span class="text">Radio Stations</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Podcasts">
              <phapplepodcastslogo class="icon" aria-hidden="true" size="24" weight="duotone"></phapplepodcastslogo><span class="text">Podcasts</span>
            </a>
            <!--v-if-->
          </li>
        </ul>
      </li>
      <li class="sidebarItem"><a class="titleLink" title="Your library">
          <phmusicnotes class="titleIcon" aria-hidden="true" weight="duotone" size="32"></phmusicnotes>
          <h4 class="title">Your library</h4>
        </a>
        <ul class="items">
          <li><a exactactiveclass="current" class="iconButton link link" title="Albums">
              <phvinylrecord class="icon" aria-hidden="true" size="24" weight="duotone"></phvinylrecord><span class="text">Albums</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Artists">
              <phusersthree class="icon" aria-hidden="true" size="24" weight="duotone"></phusersthree><span class="text">Artists</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Playlists">
              <phplaylist class="icon" aria-hidden="true" size="24" weight="duotone"></phplaylist><span class="text">Playlists</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Favourites">
              <phheart class="icon" aria-hidden="true" size="24" weight="duotone"></phheart><span class="text">Favourites</span>
            </a>
            <!--v-if-->
          </li>
          <li><a exactactiveclass="current" class="iconButton link link" title="Genres">
              <phwaveform class="icon" aria-hidden="true" size="24" weight="duotone"></phwaveform><span class="text">Genres</span>
            </a>
            <!--v-if-->
          </li>
        </ul>
      </li>
      <li class="sidebarItem"><a class="titleLink" title="Playlists">
          <phplaylist class="titleIcon" aria-hidden="true" weight="duotone" size="32"></phplaylist>
          <h4 class="title">Playlists</h4>
        </a>
        <ul class="items">
          <li><a exactactiveclass="current" class="iconButton link link" title="Random">
              <phqueue class="icon" aria-hidden="true" size="24" weight="duotone"></phqueue><span class="text">Random</span>
            </a>
            <!--v-if-->
          </li>
          <li>
            <!--v-if-->
            <!--v-if-->
          </li>
          <li>
            <!--v-if--><button class="iconButton fullWidth link link" type="button" title="Add playlist" data-test-id="item-component">
              <phlistplus class="icon" aria-hidden="true" size="24" weight="duotone"></phlistplus><span class="text"> Add playlist </span>
            </button>
          </li>
        </ul>
      </li>
    </ul>
  </nav>
</div>"
`;
